### PR TITLE
Publish cli crate

### PR DESCRIPTION
**Pull request by ry**
- Combines cli_snapshots and js into one directory.
- Extracts TS version at compile time rather than runtime
- Makes publish work on deno_cli, deno_cli_snapshots, and deno_typescript
- bumps version awkwardly. sorry.
- adds new git submodule deno_typescript/typescript
- improves lint.py and format.py scripts

Try: `cargo install deno_cli`

Towards https://github.com/denoland/deno/issues/2942

### CODE REVIEW

File: website/manual.md [hunk 1]
```diff
@@ -120,6 +120,12 @@ Using [Homebrew](https://brew.sh/) (mac):
 brew install deno
 ```
 
+To compile from source sit down with a good book and run:
```
Comment (bartlomieju) [inline website/manual.md]: 😅 it's not that bad, is it? It takes roughly 5-7 mins to compile from scratch on my machine

Comment (ry) [reply website/manual.md]: fixed text

File: cli/version.rs [hunk 2]
```diff
@@ -1,17 +1,7 @@
...
-
-pub fn typescript() -> String {
-  // TODO: By using include_str! we are including the package.json into
-  // the deno binary using serde to decode it at runtime. This is suboptimal
-  // in space and time. We need to extract the TypeScript version at compile
-  // time instead. This will be easier after #2608.
-  let data = include_str!("../node_modules/typescript/package.json");
-  let pkg: serde_json::Value = serde_json::from_str(data).unwrap();
-  pkg["version"].as_str().unwrap().to_string()
-}
```
Comment (ry) [inline cli/version.rs]: cc @kt3k

File: js/ts_global.d.ts [hunk 3]
```diff
@@ -5,7 +5,12 @@
 // module.
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-import * as ts_ from "../node_modules/typescript/lib/typescript.d.ts";
+// Warning! This is a magical import. We don't want to have multiple copies of
+// typescript.d.ts around the repo, there's already one in
+// deno_typescript/typescript/lib/typescript.d.ts. Ideally we could simply point
+// to that in this import specifier, but "cargo package" is very strict and
+// requires all files to be present in a crate's subtree.
+import * as ts_ from "$asset$/typescript.d.ts";
```
Comment (ry) [inline js/ts_global.d.ts]: cc @kitsonk

File: deno_typescript/lib.rs [hunk 4]
```diff
@@ -18,11 +18,16 @@ use std::path::PathBuf;
 use std::sync::Arc;
 use std::sync::Mutex;
 
-static TYPESCRIPT_CODE: &str =
-  include_str!("../third_party/node_modules/typescript/lib/typescript.js");
+static TYPESCRIPT_CODE: &str = include_str!("typescript/lib/typescript.js");
 static COMPILER_CODE: &str = include_str!("compiler_main.js");
 static AMD_RUNTIME_CODE: &str = include_str!("amd_runtime.js");
 
+pub fn ts_version() -> String {
```
Comment (kt3k) [inline deno_typescript/lib.rs]: 👍

File: js/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "deno_cli_snapshots"
+version = "0.18.2"
+license = "MIT"
+authors = ["Ryan Dahl <ry@tinyclouds.org>"]
```
Comment (kt3k) [inline js/Cargo.toml]: How about adding `"the Deno authors"`?

Comment (ry) [reply js/Cargo.toml]: fixed

File: js/ts_global.d.ts [hunk 6]
```diff
@@ -4,8 +4,13 @@
 // when building Deno, but the `typescript/lib/typescript.d.ts` is defined as a
 // module.
 
+// Warning! This is a magical import. We don't want to have multiple copies of
```
Comment (piscisaureus) [inline js/ts_global.d.ts]: I don't understand this comment. This is typescript code, what does cargo have to do with it?

File: tools/lint.py [hunk 7]
```diff
@@ -20,13 +20,11 @@
     "--repository=core/libdeno", "--extensions=cc,h", "--recursive", "core"
 ])
 
-run([
-    "node", eslint, "--max-warnings=0", "./js/**/*.{ts,js}",
-    "./core/**/*.{ts,js}", "./tests/**/*.{ts,js}"
-])
+run(["node", eslint, "--max-warnings=0"] +
+    find_exts2(["js", "core", "deno_typescript"], [".ts", ".js"]))
```
Comment (piscisaureus) [inline tools/lint.py]: Is there a reason to stop linting files in //tests ?

File: tools/util.py [hunk 8]
```diff
@@ -186,6 +186,14 @@ def find_exts(directories, extensions, skip=None):
     return matches
 
 
+def find_exts2(dirs, exts):
```
Comment (piscisaureus) [inline tools/util.py]: This needs a rebase after the tools clean-up that just landed.
I think it'll make life easier for you; it is much more likely to do the right thing already.

**Comment by piscisaureus** [comment]
> * Makes publish work on deno_cli, deno_cli_snapshots, and deno_typescript

I don't see any updates to cargo_package.py. Am I missing something?

File: core/Cargo.toml [hunk 9]
```diff
@@ -5,7 +5,7 @@ name = "deno"
 version = "0.18.0"
 edition = "2018"
 description = "A secure JavaScript/TypeScript runtime built with V8, Rust, and Tokio"
-authors = ["The deno authors <bertbelder@nodejs.org>"]
```
Comment (piscisaureus) [inline core/Cargo.toml]: Does crates.io accept it without an e-mail address?
Maybe we can create a shared mailbox, e.g. info@deno.land.

File: js/ts_global.d.ts (hunk 6 above)
Comment (ry) [reply js/ts_global.d.ts]: cargo package must have all dependent files present, in this subtree to work.

File: tools/util.py (hunk 8 above)
Comment (ry) [reply tools/util.py]: rebased

File: core/Cargo.toml (hunk 9 above)
Comment (ry) [reply core/Cargo.toml]: It does accept it without an email address: https://crates.io/crates/deno_cli

**Comment by ry** [comment]
> I don't see any updates to cargo_package.py. Am I missing something?

So, the other crates (deno_cli, deno_typescript, deno_cli_snapshots) are able to be published with just "cargo publish". I think we should leave to cargo_package.py as it is, because it's doing a particular hack... I could add tools/cargo_publish_others.py (or something) that simply ran cargo publish in the various directories? But I think there are more clean ups to happen still in this area, and I would just as well put that off for a while.

File: js/ts_global.d.ts (hunk 6 above)
Comment (piscisaureus) [reply js/ts_global.d.ts]: So the part I don't understand is: how does cargo even know this is a dependent file?

File: core/Cargo.toml (hunk 9 above)
Comment (piscisaureus) [reply core/Cargo.toml]: Cool, thanks

**Comment by piscisaureus** [comment]
> So, the other crates (deno_cli, deno_typescript, deno_cli_snapshots) are able to be published with just "cargo publish". I think we should leave to cargo_package.py as it is, because it's doing a particular hack... I could add tools/cargo_publish_others.py (or something) that simply ran cargo publish in the various directories? But I think there are more clean ups to happen still in this area, and I would just as well put that off for a while.

Ok, that's great. I was mainly looking for it to understand what the release procedure looks like.
Which crates have to be published, in which order? What command do I run?

**Comment by ry** [comment]
I've added `tools/cargo_publish_others.py` which shows the order they have to be run.

We should probably also be running this with the `--dry-run` argument in CI, because "cargo publish" seems to be different than "cargo build". I would leave running it in CI until later tho.

**Pull request merged**